zallet/src/components/sync.rs

```diff
@@ -1,45 +1,323 @@
+        // Ensure the wallet is in a state that the sync tasks can work with.
+        let chain = chain_view.subscribe().await?.inner();
+        let mut db_data = db.handle().await?;
+        let starting_tip = initialize(chain, db_data.as_mut()).await?;
+
+        // Spawn the ongoing sync tasks.
+        let chain = chain_view.subscribe().await?.inner();
+        let steady_state_task = tokio::spawn(async move {
+            steady_state(&chain, &params, db_data.as_mut(), starting_tip).await?;
+            Ok(())
+        });
+
+        let chain = chain_view.subscribe().await?.inner();
+        let mut db_data = db.handle().await?;
+        let recover_history_task = tokio::spawn(async move {
+            recover_history(chain, &params, db_data.as_mut(), 1000).await?;
+            Ok(())
+        });
+
+        Ok((steady_state_task, recover_history_task))
+    }
+}
+
+/// Prepares the wallet state for syncing.
+///
+/// Returns the boundary block between [`steady_state`] and [`recover_history`] syncing.
+#[tracing::instrument(skip_all)]
+async fn initialize(
+    chain: FetchServiceSubscriber,
+    db_data: &mut DbConnection,
+) -> Result<ChainBlock, SyncError> {
+    info!("Initializing wallet for syncing");
+
+    // Notify the wallet of the current subtree roots.
+    steps::update_subtree_roots(&chain, db_data).await?;
+
+    // // Fetch the wallet's current view of the chain tip.
+    // let current_tip = match (db_data.chain_height()?, db_data.block_max_scanned()?) {
+    //     // The wallet has been synced before.
+    //     (Some(tip_height), Some(max_scanned)) => {
+    //         // Fetch the block to obtain its parent's hash.
+    //         let max_scanned_block = ChainBlock::resolve(&chain, max_scanned.block_hash()).await?;
 
-            let mut interval = time::interval(Duration::from_secs(30));
+    //         if max_scanned_block.height != tip_height {
+    //             // The wallet shut down between the scan queue being updated for the chain
+    //             // tip height, and the chain tip block being scanned. We have no idea what
+    //             // block the wallet was being notified of (as reorgs could have occurred
+    //             // while the wallet was offline), so truncate it back to the max scanned
+    //             // height.
+    //             db_data.truncate_to_height(max_scanned_block.height)?;
+    //         }
+
+    //         max_scanned_block
+    //     }
+    //     // The wallet either hasn't been synced before, or hasn't scanned any blocks. In
+    //     // either case, the wallet has no state that depends on a particular block being
+    //     // present, so we can just notify it of the current chain tip.
+    //     _ => {
+    //         let current_tip = steps::get_chain_tip(&chain).await?;
+    //         info!("Latest block height is {}", current_tip.height);
+    //         db_data.update_chain_tip(current_tip.height)?;
+    //         current_tip
+    //     }
+    // };
```

Comment:
Why is this commented out?

Reply:
I was just in the middle of fixing this TODO. It's done now.